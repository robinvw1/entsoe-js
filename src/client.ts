import xml2js from 'xml2js';
import { ofetch } from 'ofetch';
import { DateTime } from 'luxon';

// Constants
import { Endpoint, DocumentType } from '@/constants';

// Types
import type { ClientOptions, DayAheadPricesOptions } from '@/types';

// Transformer
import { dayAheadPricesSchema, transformResponse, DayAheadPricesResponse } from '@/transformer';

export class Client {
    private readonly client;
    private readonly parser;
    private readonly securityToken;

    constructor(options: ClientOptions) {
        if (!options || typeof options !== 'object') {
            throw new Error('options must be an object');
        }

        const { securityToken } = options;

        if (!securityToken || !securityToken.length) {
            throw new Error('options.securityToken is missing');
        }

        this.securityToken = securityToken;
        this.parser = new xml2js.Parser();
        this.client = ofetch.create({
            baseURL: options.endpoint || Endpoint.Base,
            headers: {
                'Content-Type': 'application/xml'
            },
            onRequest: ({ options }) => {
                /* c8 ignore next */
                options.query = options.query || {};
                options.query.securityToken = this.securityToken;
            },
            parseResponse: async (responseText) => {
                return await this.parser.parseStringPromise(responseText);
            }
        });
    }

    public async dayAheadPrices(options: DayAheadPricesOptions) {
        if (!options || typeof options !== 'object') {
            throw new Error('options must be an object');
        }

        const { biddingZone, startDate, endDate, documentType = DocumentType.PriceDocument } = options;

        if (!biddingZone || !biddingZone.length) {
            throw new Error('options.biddingZone is missing');
        }

        const now = DateTime.utc();
        const sd = startDate ? DateTime.fromISO(startDate) : now.startOf('day');
        const ed = endDate ? DateTime.fromISO(endDate) : now.startOf('day').plus({ day: 1 });
        const timeInterval = `${sd.toUTC().toISO({ suppressSeconds: true })}/${ed.toUTC().toISO({ suppressSeconds: true })}`;

        const response = await this.client<DayAheadPricesResponse>('/api', {
            query: {
                in_Domain: biddingZone,
                out_Domain: biddingZone,
                timeInterval,
                documentType
            }
        });

        const validatedResponse = dayAheadPricesSchema.parse(response); // TODO: maybe not on runtime
        const parsedDayAheadPrices = transformResponse(validatedResponse);

        return parsedDayAheadPrices;
    }
}
