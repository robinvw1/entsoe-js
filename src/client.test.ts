import { describe, expect, it } from 'vitest';
import { Client } from './client';
import { BiddingZone } from './constants';

describe('client', () => {
    it('should instantiate', () => {
        const client = new Client({ securityToken: 'foo' });

        expect(client).toBeInstanceOf(Client);
    });

    it('should throw on missing options', () => {
        // @ts-expect-error options missing
        expect(() => new Client()).toThrowError('options must be an object');
        expect(() => new Client({ securityToken: '' })).toThrowError('options.securityToken is missing');
    });
});

describe('dayAheadPrices', () => {
    const client = new Client({ securityToken: 'bar' });

    it('should throw on missing options', async () => {
        // @ts-expect-error options missing
        await expect(client.dayAheadPrices()).rejects.toThrowError('options must be an object');
        // @ts-expect-error options incorrect
        await expect(client.dayAheadPrices({ biddingZone: '' })).rejects.toThrowError('options.biddingZone is missing');
    });

    it('should throw on invalid securityToken', async () => {
        const client = new Client({ securityToken: 'invalid' });

        await expect(client.dayAheadPrices({ biddingZone: BiddingZone.NL })).rejects.toThrowError('401 Unauthorized');
    });

    it('should handle bad requests', async () => {
        const invalidDate = '2020-01-01T00:00Z';

        await expect(
            client.dayAheadPrices({
                biddingZone: BiddingZone.NL,
                startDate: invalidDate,
                endDate: invalidDate
            })
        ).rejects.toThrowError('400 Bad Request');
    });

    it('should match snapshot', async () => {
        const response = await client.dayAheadPrices({ biddingZone: BiddingZone.NL });

        expect(response).toMatchSnapshot();
    });

    it('should handle custom startDate and endDate', async () => {
        const response = await client.dayAheadPrices({
            biddingZone: BiddingZone.UK,
            startDate: '2020-01-01T00:00:00Z',
            endDate: '2020-01-10T00:00:00Z'
        });

        expect(response[0].period_start).toBe('2020-01-01T00:00Z');
        expect(response[0].period_end).toBe('2020-01-02T00:00Z');

        expect(response[0].records[0].timestamp).toBe('2020-01-01T00:00Z');

        expect(response.at(-1)?.period_start).toBe('2020-01-09T00:00Z');
        expect(response.at(-1)?.period_end).toBe('2020-01-10T00:00Z');
    });
});
