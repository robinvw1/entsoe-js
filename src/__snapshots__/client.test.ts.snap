// Vitest Snapshot v1

exports[`dayAheadPrices > should match snapshot 1`] = `
[
  {
    "currency": "EUR",
    "period_end": "2022-11-22T23:00Z",
    "period_start": "2022-11-21T23:00Z",
    "records": [
      {
        "price": "112.42",
        "timestamp": "2022-11-21T23:00Z",
      },
      {
        "price": "111.84",
        "timestamp": "2022-11-22T00:00Z",
      },
      {
        "price": "99.97",
        "timestamp": "2022-11-22T01:00Z",
      },
      {
        "price": "82.11",
        "timestamp": "2022-11-22T02:00Z",
      },
      {
        "price": "87.65",
        "timestamp": "2022-11-22T03:00Z",
      },
      {
        "price": "98.34",
        "timestamp": "2022-11-22T04:00Z",
      },
      {
        "price": "148.04",
        "timestamp": "2022-11-22T05:00Z",
      },
      {
        "price": "219.05",
        "timestamp": "2022-11-22T06:00Z",
      },
      {
        "price": "248.97",
        "timestamp": "2022-11-22T07:00Z",
      },
      {
        "price": "236.90",
        "timestamp": "2022-11-22T08:00Z",
      },
      {
        "price": "242.20",
        "timestamp": "2022-11-22T09:00Z",
      },
      {
        "price": "244.58",
        "timestamp": "2022-11-22T10:00Z",
      },
      {
        "price": "245.32",
        "timestamp": "2022-11-22T11:00Z",
      },
      {
        "price": "251.76",
        "timestamp": "2022-11-22T12:00Z",
      },
      {
        "price": "258.96",
        "timestamp": "2022-11-22T13:00Z",
      },
      {
        "price": "260.00",
        "timestamp": "2022-11-22T14:00Z",
      },
      {
        "price": "255.61",
        "timestamp": "2022-11-22T15:00Z",
      },
      {
        "price": "284.96",
        "timestamp": "2022-11-22T16:00Z",
      },
      {
        "price": "294.90",
        "timestamp": "2022-11-22T17:00Z",
      },
      {
        "price": "281.53",
        "timestamp": "2022-11-22T18:00Z",
      },
      {
        "price": "254.20",
        "timestamp": "2022-11-22T19:00Z",
      },
      {
        "price": "246.54",
        "timestamp": "2022-11-22T20:00Z",
      },
      {
        "price": "215.00",
        "timestamp": "2022-11-22T21:00Z",
      },
      {
        "price": "171.50",
        "timestamp": "2022-11-22T22:00Z",
      },
    ],
    "unit": "MWH",
  },
]
`;
